Add external krb5 test support

frozencemetery opened the pull request:
##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed ~(I think this change is trivial)~

##### Description of change
As suggested by @mattcaswell and @tiran in https://github.com/openssl/openssl/issues/1903, this PR augments the external tests with [krb5](https://github.com/krb5/krb5)'s test suite (built against openssl).  I am a krb5, not an OpenSSL, developer, so mistakes are expected on my part.  In particular, I do not know Perl, and as such am not wedded to this design in any particular way.

Thanks!

levitte commented on test/recipes/90-test_external.t:
```diff
@@ -11,34 +11,61 @@ use OpenSSL::Test;
+                           "./configure --with-ldap --with-prng-alg=os " .
+                           "--with-crypto-impl=openssl --enable-pkinit " .
+                           "--with-pkinit-crypto-impl=openssl " .
+                           "--with-tls-impl=openssl && " .
```
Isn't this making a huge assumption about where the openssl libs and includes are located?

kaduk replied:
Yup.  And the krb5 configure doesn't seem to have logic specifically to find an openssl at a path, so I think it's the manual CPPFLAGS/LDFLAGS route for us.

levitte replied:
CFLAGS and LDFLAGS it is, yes.  You could set the values like this:

``` perl
    "CFLAGS='-I" . bldtop_dir("include"). " -I" . srctop_dir("include") . "' " .
    "LDFLAGS='-L" . bldtop_dir() . "' " .
```

kroeckx commented:
It should probably change the travis configuration so that this actually gets tested.

levitte commented:
Hi, this got lost for a bit.  I hope you're still with us, @frozencemetery

frozencemetery commented:
Yes, sorry.  I had to deprioritize this due to krb5-1.15 release, but should have a new version in the next week.

levitte commented:
One more thing.  I had a look at our CLA db, and can't find any entry for you, so we need you to sign an ICLA.

See https://www.openssl.org/policies/cla.html for more information

frozencemetery commented:
I was hoping this change would be small enough to not need CLA... I guess not.  I'll look into it.

richsalz commented:
If you think it is trivial, push an updated commit that adds "CLA: trivial" to the text.

levitte commented:
I don't see this as a trivial change, for the record.

mattcaswell commented:
No, this is definitely not trivial.

richsalz commented:
Okay, then @frozencemetery (cool login name BTW), we'll need a CLA, or you can close this.

frozencemetery commented:
Thanks!  Emailed signed CLA.

richsalz commented:
Got your CLA.  But if this is coming from RedHat, we need CLA from them.  Or an individual submission, we need a non-RedHat email address.

mattcaswell commented:
We already have a RedHat CCLA. Just need Tomas Mraz to add you to schedule A.

t8m commented:
What does it mean "add you to schedule A"? Is it sufficient to send plain text e-mail to legal@opensslfoundation.org stating that we added rhardwood to schedule A?

richsalz commented:
Ideally, the person who signed the original CCLA, would send an updated "Schedule A" that we would put in the files.  The "Schedule" is an appendix, and lists the people covered by the CCLA.  If you can't do that, please contact legal@opensslfoundation.com

frozencemetery commented:
Thanks for your patience and assistance while CLA was sorted.

I have rebased against master, and updated our approach to reflect the newly added PYCA tests.

levitte commented on test/README.external:
```diff
@@ -93,8 +93,42 @@ to be installed.
+krb5's tests will then be run as part of the rest of the suite, or can be
+explicitly run (with more debugging):
+
+$ HARNESS_VERBOSE=yes VERBOSE=1 make TESTS=test_external_krb5 test
```
`HARNESS_VERBOSE=yes` is redundant, remove it please

levitte commented on test/README.external:
```diff
@@ -93,8 +93,42 @@ to be installed.
+$ HARNESS_VERBOSE=yes VERBOSE=1 make TESTS=test_external_krb5 test
+
+Test failures supressions
+----------------
```
The line of dashes should be the same length as the title it underlines...

levitte commented:
I wonder if there should be something similar for heimdal...

mattcaswell commented on test/recipes/95-test_external_krb5.t:
```diff
@@ -0,0 +1,21 @@
+
+plan tests => 1;
+
+ok(run(cmd([data_file("krb5.sh")])), "running krb5 tests");
```
We should skip this if krb5 is not available. Similar to the skip we have in 95-test_external_pyca.t

mattcaswell commented on test/recipes/95-test_external_krb5.t:
```diff
@@ -0,0 +1,21 @@
+#! /usr/bin/env perl
+# Copyright 2015-2017 The OpenSSL Project Authors. All Rights Reserved.
```
Just 2017

frozencemetery commented:
Thanks, updated to fix.

frozencemetery commented:
Regarding Heimdal - I'm not involved in that project so I'm not the right person to implement that if it should exist.  Until very recently (and possibly still) Heimdal had serious issues building on 32-bit platforms, which is fine for Travis (x86_64) but worth keeping in mind for any test code.

levitte commented:
I had a look at Heimdal earlier today, and it seems they have made themselves independent of OpenSSL, basically reimplementing *everything* they needed (and probably more, I found a reimplementation of the ENGINE interface among other things)

levitte commented:
... so basically, I don't think we need to bother with Heimdal at all...

kaduk commented:
Fixing the 32-bit issues was part of the work done in the lead up to the recent Heimdal release (the first in five years).
I don't think it's accurate to say that Heimdal has completely moved off of OpenSSL; they still have lots of code to (optionally) use OpenSSL for libcrypto, though they also have a bundled libhcrypto that may be more commonly used.  (MIT krb5 also has a bundled crypto library, though, but requires OpenSSL for certain PKINIT-related functionality.)
So, it may be worth having a Heimdal external test, if someone feels like putting the time into it.  (I am not likely to volunteer.)

frozencemetery commented:
Hi, anything else I can do for this?  @mattcaswell, have your changes been addressed?  Thanks!

kaduk commented on test/recipes/95-test_external_krb5_data/krb5.sh:
```diff
@@ -0,0 +1,23 @@
+# https://www.openssl.org/source/license.html
+
+# krb5's test suite clears LD_LIBRARY_PATH
+LDFLAGS="-L`pwd`/$BLDTOP -Wl,-rpath,`pwd`/$BLDTOP"
```
For some reason I am remembering a discussion of new-dtags vs. -rpath, but didn't see it in a skim of the overview page.
In any case, I guess it doesn't really matter here, since this is just for running the test suite.

levitte commented:
Merged.  7c82f5673bea290fac52b2d41932f6125f56802c to 0fef74486ede58e34eeac2fab38c3f467ea31f95
